### ISSUE TITLE
Multi GPU

### ISSUE THREAD

**Issue by borgr** [davda54/sam#61]
IIUC combining the two answers should work and probably is a better example (easier for people to change their code because it is pytorch native). Didn't want to pull request in case I am mistaken

for input, output in data:
  def closure():
    loss = loss_function(output, model(input))
    loss.backward()
    return loss

  loss = loss_function(output, model(input))
  with model.no_sync(): # << the important change is here
    loss.backward()
  optimizer.step(closure)
  optimizer.zero_grad()

**Comment by borgr** [comment]
This is important not only for convenience but also for things that wrap the optimizer (e.g. huggingface accelerator) and expect a certain API

**Comment by davda54** [comment]
Hi, thanks for your suggestion! If I am not mistaken this only works for `DistributedDataParallel`, right? Not as good for a simple introductory example if it doesn't work for a general `nn.Module`.
